### ISSUE TITLE
DeprecationWarning: "'cgi' is deprecated and slated for removal in Python 3.13"

### ISSUE THREAD

**Issue by k-bu** [twisted/twisted#11848]
**Describe the incorrect behavior you saw**
The currently used package "cgi" will be removed in Python 3.13.

**Describe how to cause this behavior**
```
> python -W always
Python 3.11.0 (main, Feb  3 2023, 12:28:31) [MSC v.1929 64 bit (AMD64)] on win32
>>> from twisted.web import http
...\site-packages\twisted\web\http.py:103: DeprecationWarning: 'cgi' is deprecated and slated for removal in Python 3.13
  import cgi
```

**Describe the correct behavior you'd like to see**
No warning.

**Testing environment**
 - Windows: 11
 - Python: 3.11.0
 - Twisted: 22.10.0

**Comment by adiroiban** [comment]
Thanks for the report. I am not sure what is the bug here :)

I see that you are using 3.11 .
I guess that the `cgi` module is still present and Twisted works as expected.

I think that this is more of a refactoring task / future-proof task, rather than a bug.

Is there something broken with Twisted?

Happy to review a PR in which the `cgi` module is removed and replaced with something else.

**Comment by k-bu** [comment]
Yes sorry, it's not a bug. Everything's working fine! I just didn't know how to mark it as a refactoring task afterwards.

**Comment by befeleme** [comment]
Hello,
we've  started building Fedora packages with Python 3.13.0 alphas and there's quite a lot failures when building twisted. 

```
FAILED (skips=640, failures=14, errors=51, successes=9085)
```

They're not only related to the removed `cgi`, see full build logs: https://download.copr.fedorainfracloud.org/results/@python/python3.13/fedora-rawhide-x86_64/06770350-python-twisted/builder-live.log.gz

**Comment by adiroiban** [comment]
Many thanks Karolina for the report.

The `cgi` issue should already be fixed in `trunk` ... but not yet released 
This was done in https://github.com/twisted/twisted/issues/12015

Also, the skip test `unittest` part was already fixed in `trunk` as part of #12054

-------

Since `cgi` is fixed , I am going to close this and we can continue reporting python 3.13 errors over #11750 and try to fix the issues in separate/dedicated tickets.

**Comment by hroncok** [comment]
> Since `cgi` is fixed , I am going to close this...

I see a cgi import in:

https://github.com/twisted/twisted/blob/c0c6440827c47797415b6ddf5bbf0b60a203da02/src/twisted/web/http.py#L104

This is in trunk.

**Comment by adiroiban** [comment]
True. Many thanks.

It is used here https://github.com/twisted/twisted/blob/c0c6440827c47797415b6ddf5bbf0b60a203da02/src/twisted/web/http.py#L991

We should fix it.

**Comment by adiroiban** [comment]
PR ready for review https://github.com/twisted/twisted/pull/12058

**Comment by defnull** [comment]
I know it's too late as #12058 was already merged and you all need a quick solution for Python 3.13, but please re-evaluate the use of `email.message` in this context. Multipart as used in emails is not the same as [multipart/form-data](https://datatracker.ietf.org/doc/html/rfc7578) as used by browsers. RFC-7578 is way simpler and stricter than what needs to be supported by email parsers. A web framework should not accept nested multipart streams, segments without a name, filenames with invalid character in them, very CPU intensive but useless content encodings, broken input that is common for email but absolutely not allowed for browsers, and other stuff that no browser or HTTP client should ever emit. The default policy for `email.message_from_bytes` is  to be as compatible as possible. When dealing with untrusted user input, stricter is usually better. This might actually be a security risk.

I wrote https://pypi.org/project/multipart/ years ago because `cgi.FieldStorage` was buggy and the `email` stdlib module is way to complex for just parsing `multipart/form-data`. It is now part of [bottle](https://github.com/bottlepy/bottle) and [treq](https://github.com/twisted/treq) also adopted it recently. There is a new release with a new non-blocking and very fast parser, which also enforces strict limits to prevent a large group of resource exhaustion attacks using manipulated POST requests. Maybe give it a try?

**Comment by glyph** [comment]
> I wrote https://pypi.org/project/multipart/ years ago because `cgi.FieldStorage` was buggy and the `email` stdlib module is way to complex for just parsing `multipart/form-data`. It is now part of [bottle](https://github.com/bottlepy/bottle) and [treq](https://github.com/twisted/treq) also adopted it recently.

I should note that we cannot depend on it directly as such a dependency precludes Twisted users from using the quite popular Starlette framework.  Until the two frameworks can settle the naming dispute somehow, we would need to vendor it into Twisted and then *also* vendor it into Treq, which is very unfortunate.

**Comment by adiroiban** [comment]
I am more than happy to review and accept a PR with multipart.

For my production code, I am not using the Twisted parser ... as I need a streaming / non-blocking parser... and I am using a forked Twised version, as upstream Twisted has not support so streaming :(

multipart looks like a good fit.

-------------

>  A web framework should not accept nested multipart streams, segments without a name, filenames with invalid character in them, very CPU intensive but useless content encodings, broken input that is common for email but absolutely not allowed for browsers, and other stuff that no browser or HTTP client should ever emit.

The email.message was good enough to pass the existing automated tests.

It would be nice to have a PR that adds the tests in which nested multiparts are rejected.

Also, some benchmark tests would be awesome.

**Comment by defnull** [comment]
The name conflict with `python-multipart` (discussed [here](https://github.com/twisted/treq/issues/399) and [here](https://github.com/defnull/multipart/issues/53#issuecomment-2362099457)) is really a bummer :/

As for the benchmarks, I ran some quick tests yesterday, parsing a multipart stream with a single 16MB file using the non-blocking and blocking APIs (if available). The parsers were configured with default settings, the results were thrown away.

| Implementation  | Blocking (ramdisk) | Non-Blocking |
| ------------- | ------------- | ------------- |
| multipart  | 767.14 MB/s | 3014.40 MB/s  |
| werkzeug  | 686.90 MB/s  |  2164.95 MB/s |
| python-multipart  |  404.18 MB/s | 611.95 MB/s  |
| (stdlib) email  |  183.05 MB/s | not available |

(Best results after 10*3 rounds each, higher throughput is better)

It is also very important to test edge cases (e.g. a file that consists only of `\r\n` or parts of the boundary) because this can cause an extreme slowdown in some parsers. I measured less than 1MB/s for one of them, which could qualify as a DoS vector.

**Comment by adiroiban** [comment]
>  I ran some quick tests yesterday, parsing a multipart stream with a single 16MB file using the non-blocking and blocking APIs (if available).

Thanks for looking into this.

Can you share the code?

It would help to create some benchmark tests for Twisted using the existing stdlib code.
Once we migrate to `multiplart` we can observe the changes.

--------

> The name conflict with python-multipart (discussed here and here) is really a bummer :/

Well... `multipart` is kind of generic.
When I read `multipart` I am thinking at https://datatracker.ietf.org/doc/html/rfc1341

Maybe, you can rename the project to something like `multiplart_formdata` of maybe `rfc7578` to make it clear that this is only about a specific multipart handling.

-------

For my personal projects I need to handle AS2 payloads... so I will continue to use a custom parser... and a fork of Twisted, as for example I need support for [100-Continue](https://github.com/twisted/twisted/issues/6928)

----------

Another option is to deprecate multipart support in Twisted.

If people need to handle multipart content, they can use treq,  or use their preferred library.

**Comment by glyph** [comment]
> The name conflict with `python-multipart` (discussed [here](https://github.com/twisted/treq/issues/399) and [here](https://github.com/defnull/multipart/issues/53#issuecomment-2362099457)) is really a bummer :/

Yeah :(  It's definitely a flaw in the ecosystem that we have no way to appeal this, and that if both maintainers say "no" there's not really a path forward. (I have my own opinion about which maintainer is "correct" here, but it doesn't much matter in terms of consequences!)

**Comment by defnull** [comment]
> Can you share the code?

I'm working on a proper benchmark suite, but at the moment some parsers show strange behavior and I'd like to sort those out before publishing anything and call it a "benchmark" and not just "quick tests". For example: After rewriting the benchmark a bit, werkzeug now shows way worse performance. That can't be right. And `python-multipart` has two catastrophic edge-cases that are already reported and need a fix. Triggering worst-case scenarios would not be fair in a benchmark.

tl;dr; not yet, needs work

**Comment by defnull** [comment]
--> https://github.com/defnull/multipart_bench

**Comment by adiroiban** [comment]
Many thanks for putting together the benchmarks.

I am happy to review a PR that uses werkzeug or multipart .

I am more +1 for werkzeug .

My question is: why not improve the multipart support in werkzeug, instead of create yet another project

Everyone is free to work on whatever they want in the free time.
So I am not judging.

But in the spirit of colaboration, to me it make sense to use werkzeug in an atempt to try to reivent the wheel

**Comment by defnull** [comment]
> My question is: why not improve the multipart support in werkzeug, instead of create yet another project

I had several reasons 14 years ago and some of them still hold today.
* Most importantly, werkzeug does a lot more than just multipart parsing. It's a huge library with 20k lines of code and tons of functionality. `multipart.py` does exactly one thing and is 95% smaller (less than 1k lines). It is also easier to understand, debug and secure because of that.
* Multipart is significantly faster. Werkzeug is still fast (compared to the others) but multipart beats it in benchmarks by a significant margin. Yes, web framework benchmarks are sometimes questionable. A couple of milliseconds per request do not really matter. But when dealing with potentially huge file uploads, that's a different story.
* There is one other very significant issue with werkzeug that I do not want to disclose publicly until it's fixed.

**Comment by adiroiban** [comment]
Thanks for the feedback.

I am not -1 for using `multipart` in Twisted and I happy to approve and review a PR for Twisted using `multipart`.

The 20k of extra lines of code, should not be a big deal for Twisted.
In Twisted we already include a ton of lines that might not be used by everyone ... legacy FTP, IRC, jabber, some NMEA gps code

**Comment by defnull** [comment]
I found a bug in my test cases for `email` and `python-multipart` fixed their critical performance issue, so I updated the benchmark with up-to-date numbers. Hope the benchmark discussion is not too off-topic here.

As for the PR integrating multipart, there is a blocker I'd rather not discuss in public. That should be resolved first. 

Edit: Turns out the blocking issue I mentioned is [known since 2009](https://github.com/twisted/twisted/issues/4688#issuecomment-2385711609) and its security implications were already discussed in public, so no need for secrecy.

**Comment by twm** [comment]
FWIW I am a strong -1 on Werkzeug. It is a fine library but I don't think that it is suitable as a Twisted dependency because we have a different concept of backwards-compatibility. Using Werekzeug could result in dependency hell for WSGI applications hosted using Twisted.

**Comment by defnull** [comment]
Quick update: The `python-multipart` project found a way to transition to a new import name and avoid the name collision with `multipart` in the future. The transition is not complete yet, but we are making good progress.

**Comment by glyph** [comment]
That's fantastic news!
